python/ray/air/result.py

```diff
@@ -110,3 +117,109 @@ def _repr(self, indent: int = 0) -> str:
+
+        Returns:
+            A :py:class:`Result` object of that trial.
+        """
+
+        # TODO(yunxuanx): restoration from cloud storage
+        local_path = path
+
+        cls._validate_trial_dir(local_path)
+
+        # Restore reported metrics
+        with open(Path(local_path) / EXPR_RESULT_FILE, "r") as f:
+            json_list = [json.loads(line) for line in f if line]
+            metrics_df = pd.json_normalize(json_list, sep="/")
+
+        metrics = metrics_df.iloc[-1].to_dict() if not metrics_df.empty else {}
+
+        # Restore all checkpoints from the checkpoint folders
+        ckpt_dirs = [
+            os.path.join(local_path, entry)
+            for entry in os.listdir(local_path)
+            if entry.startswith("checkpoint_")
+        ]
+
+        if ckpt_dirs:
+            checkpoints = [
+                Checkpoint.from_directory(ckpt_dir) for ckpt_dir in ckpt_dirs
+            ]
+
+            checkpoint_metrics = [
+                metrics_df[metrics_df[TRAINING_ITERATION] == ckpt.iteration].to_dict(
+                    "records"
+                )[0]
+                for ckpt in checkpoints
+            ]
+
+            # TODO(air-team): make metrics a property of Checkpoint
+            best_checkpoints = list(zip(checkpoints, checkpoint_metrics))
+            latest_checkpoint = max(checkpoints, key=lambda ckpt: ckpt.id)
+        else:
+            best_checkpoints = latest_checkpoint = None
+
+        # Restore the trial error if it exists
+        error = None
+        error_file_path = Path(local_path) / EXPR_ERROR_PICKLE_FILE
+        if os.path.exists(error_file_path):
+            error = pickle.load(open(error_file_path, "rb"))
+
+        return Result(
+            metrics=metrics,
+            checkpoint=latest_checkpoint,
+            _local_path=local_path,
+            _remote_path=None,
+            metrics_dataframe=metrics_df,
+            best_checkpoints=best_checkpoints,
+            error=error,
+        )
+
+    @PublicAPI(stability="alpha")
+    def get_best_checkpoint(self, metric: str, mode: str) -> Optional[Checkpoint]:
+        """Gets best persistent checkpoint of this trial.
+
+        Any checkpoints without an associated metric value will be filtered out.
+
+        Args:
+            metric: The key for checkpoints to order on.
+            mode: One of ["min", "max"].
+
+        Returns:
+            :class:`Checkpoint <ray.air.Checkpoint>` object, or None if there is
+            no valid checkpoint associated with the metric.
+        """
+        assert self.best_checkpoints, "No checkpoint exists in the trial directory!"
+
+        assert mode in [
+            "max",
+            "min",
+        ], f'Unsupported mode: {mode}. Please choose from ["min", "max"]!'
+
+        op = max if mode == "max" else min
+        valid_checkpoints = [
+            ckpt_info for ckpt_info in self.best_checkpoints if metric in ckpt_info[1]
+        ]
```

Comment:
If there are no valid checkpoints, I think we should throw a warning and give the users the set of metric names that can be used. Then, switch this method to always return type to non-optional.

Reply:
If no valid checkpoints, we can only return None? so it has to be non-optional?

Reply:
Got it, that makes a lot of sense.